Pf2 analysis

MrBones1102 opened the pull request:
Expansion of pf2 analysis to include PaCMAP of weighted projections to see how diseases might be related.

aarmey commented on mrsa_ca_rna/factorization.py:
```diff
@@ -110,6 +111,12 @@ def perform_parafac2(data: xr.Dataset, rank: int = 10):
     rec_errors = cp.asnumpy(cp.array(rec_errors))
 
+    if l1 is not None:
+        l1_penalty = mcp.penalties.L1Penalty(l1)
```
Made an issue about this. Note that the l1 sparsity should only be applied to the genes factor matrix.

aarmey commented on mrsa_ca_rna/figures/figure13.py:
```diff
@@ -0,0 +1,70 @@
+"""File plots the pacmap of the projections and weighted projections of the pf2
```
See my comments on the other PR about renaming the figure files.

aarmey commented on mrsa_ca_rna/figures/figure13.py:
```diff
@@ -0,0 +1,70 @@
+    weighted_projections = [x @ disease_factors[1] for x in disease_projections]
+
+    mapped_p = perform_pacmap(disease_projections)
+    mapped_wp = perform_pacmap(weighted_projections)
```
You only embed the projections, not the weighted projections.

aarmey commented on mrsa_ca_rna/map.py:
```diff
@@ -0,0 +1,33 @@
+# local module imports
+
+
+def perform_pacmap(matrix_list: list):
```
No need for a function here. This is only wrapping ~2-3 lines of code.

aarmey replied:
In fact, I would just do this inside the Pf2 function.

MrBones1102 commented:
I made a number of changes which I'd like to enumerate here for clarity.

- PacMAP changes:
 - Moved pacmap to the pf2 function
 - Altered the mapping to make weighted projections (I think this is correct now but might not be)
 - Deleted `map.py`
- Regularization changes:
 - specification of l1 regularization now invokes `parafac2_aoadmm()`
  - this was part of a previous commit that I forgot to push

I would like to get `parafac2_aoadmm()` running on the GPU. I do not see a way to get matcouply running through the GPU in a way that's analogous to Tensorly, just by reading through their docs. Am I missing something or do we need to try something like cuPyNumeric? @aarmey

MrBones1102 commented:
Pyright errors need resolution. Something about the way I'm doing the returns on `perform_parafac2()`. I'm getting tangled up in the types, with tuples of tuples of lists being returned. Working through standardizing the output better to make things clearer.

aarmey commented:
Going to make some changes right now.

aarmey reviewed (changes requested):
@MrBones1102 I simplified the factorization function. You can now adjust the figure files to match.
